indexer/Indexer.h

```diff
@@ -179,18 +182,20 @@ struct PartialDocument {
 class TuIndexer final {
```

Comment:
I'm struggling a bit on what to write here. There is exactly one place where this type is constructed and used.

https://sourcegraph.com/github.com/sourcegraph/scip-clang/-/blob/indexer/Worker.cc?L795-804

How about something like:

```
// All reference fields are expected to be alive for AST visitation.
//
// Once AST visitation is complete, any data from the SymbolFormatter
// or documentMap is extracted.
```

?

I'll add this (or other suggested docs) in a follow-up PR to not block that from landing this one.

Reply:
Eh, "all reference fields are expected to be alive for AST visitation" sounds self-evident. If I was dropped in the middle of a C++ codebase my second question would be, "what's the correct way to use this abstraction?" and a big part of that is what the lifetimes of things are. (First question being, "what is this for?")

Reply:
How about something like:

```
// Helper struct to separate core indexing logic from boilerplate
// code present for Clang API impedance matching.
//
// Most fields are borrowed from the parent context, with a SymbolFormatter
// functioning as a cache+builder for symbol strings.
//
// At the end, `emitDocumentOccurrencesAndSymbols` is called
// to extract indexing data, after which this struct can be destroyed.
```

(Still not super happy with this)

There's essentially one "top-level lifetime" for this code, which is the lifetime for visitation. Everything is expected to be alive for that lifetime. Once that completes, all the data is extracted (transformed+copied) into a `scip::Index`, at which point the references are no longer needed.

With more complicated lifetimes, I would write down maybe a list of lifetimes which nest, or a tree or something. In this case, `TuIndexer` is present for separating the core indexing logic from the largely boilerplate code present for Clang API impedance matching. So everything is borrowed from the parent context, there is some computation + a final extraction step, and the `TuIndexer` is destroyed. So it seems like the level of complexity is low enough that one can understand lifetimes based on the single point where the TuIndexer type is constructed (that said, I'm obviously biased, because I wrote said code).

That's why I'm struggling to write a useful comment about lifetimes here. I understand your point in the general case, especially when lifetimes are more complex. If there were different competing lifetimes here (e.g. in a server, you'd have per-request lifetimes, which are important to keep track off vs longer lifetimes), then it'd be clearer to me what do write down...